src/components/WfsLayer.tsx

```diff
@@ -0,0 +1,103 @@
+import React, { useEffect, useState } from 'react'
```

Comment:
There is a lot of logic duplicated here that already exists in the `GeoJSONLayer` component. I think we can move this logic there to make it more re-usable. Especially the following things:

- GeoJSONLayer should respond to changes in the URL passed to it
- GeoJSONLayer should remove and update the layers when a new request complete

I think if we make these changes and add render the GeoJSONLayer based in the result of `isVisible` this should work and provide a more re-usable solution.

Reply:
I tried refactoring this locally and it cleans up the component quite nicely.

```ts
import { useMapInstance } from '@datapunt/react-maps'
import { FeatureCollection } from 'geojson'
import { GeoJSONOptions, Map } from 'leaflet'
import React, { useEffect, useState } from 'react'
import { MAX_ZOOM_LEVEL, MIN_ZOOM_LEVEL } from '../constants'
import { ZoomLevel } from '../types'
import getBBox from '../utils/getBBox'
import GeoJSONLayer from './GeoJSONLayer'

interface Props {
  url: string
  options?: GeoJSONOptions
  zoomLevel: ZoomLevel
}

const isVisible = (mapInstance: Map, zoomLevel: ZoomLevel) => {
  const { min, max } = zoomLevel
  const zoom = mapInstance.getZoom()
  return zoom <= (min || MIN_ZOOM_LEVEL) && zoom >= (max || MAX_ZOOM_LEVEL)
}

const WfsLayer: React.FC<Props> = ({ url, options, zoomLevel }) => {
  const mapInstance = useMapInstance()
  const [bbox, setBbox] = useState('')

  useEffect(() => {
    function onMoveEnd() {
      if (mapInstance) {
        setBbox(getBBox(mapInstance))
      }
    }

    mapInstance?.on('moveend', onMoveEnd)

    return () => { mapInstance?.off('moveend', onMoveEnd) }
  }, [mapInstance])

  if (!mapInstance || !isVisible(mapInstance, zoomLevel)) {
    return null
  }

  const params = bbox || getBBox(mapInstance)

  return <GeoJSONLayer url={`${url}${params}`} options={options} />
}

export default WfsLayer

```

And this only changes the implementation of `GeoJSONLayer` minimally.

```ts
import { GeoJSON, useMapInstance } from '@datapunt/react-maps'
import { GeoJsonObject } from 'geojson'
import { GeoJSON as LeafletGeoJSON, GeoJSONOptions } from 'leaflet'
import React, { useEffect, useState } from 'react'
import fetchWithAbort from '../utils/fetchWithAbort'

type Props = {
  url: string
  options?: GeoJSONOptions
}

const GeoJSONLayer: React.FC<Props> = ({ url, options }) => {
  const mapInstance = useMapInstance()
  const [layerInstance, setLayerInstance]  = useState<LeafletGeoJSON>()
  const [json, setJson] = useState<GeoJsonObject>()

  useEffect(() => {
    if (!mapInstance) {
      return
    }

    const [request, controller] = fetchWithAbort(url)

    request
      .then(res => res.json())
      .then(data => {
        layerInstance?.clearLayers()
        layerInstance?.addData(data)
        setJson(data)
      })
      .catch(error => {
        // Ignore abort errors since they are expected to happen.
        if (error instanceof Error && error.name === 'AbortError') {
          return
        }

        return Promise.reject(error)
      })

    return () => {
      controller.abort()
    }
  }, [mapInstance, url])

  return json ? <GeoJSON setInstance={setLayerInstance} args={[json]} options={options} /> : null
}

export default GeoJSONLayer
```

Reply:
Interesting solution. Unfortunately if we do that we'll introduce a memory leak while the wfs layer gets recreated for each visibility change and the layer data is not released. To solve that would mean to move almost all the functionality and logic of WfsLayer to GeoJSONLayer.  This would implement 2 completely different business logic scenarios in one component which I'm not happy to do. The GeoJSONLayer and WfsLayer are 2 concepts that serve 2 different purposes, why don't we let them for what they are? From my point of view the only common stuff of the two is that they are calling the same fetch function (but at different moments with different logic).

Reply:
I don't quite see how this would cause a memory leak, even if the layer does get re-created. This also happens in the original code, simply without the additional abstraction. The lifecycle of the component and hooks should make sure that any objects or listeners are cleaned up.

I am aware that this will move a lot of logic from the `WfsLayer` to the `GeoJSONLayer`. However I believe that this functionality certainly does belong in the `GeoJSONLayer` as it's responsibility is to retrieve GeoJSON from a URL and present it on the layer. If that URL changes so should the content displayed on the layer. It does not make sense to me to duplicate this code if we can re-use it.

Reply:
I looked again at this and I don't see the advantage of forcing two scenarios into one, and creating extra dependencies. Relating these 2 components by 'Inheritance' wouldn't simplify the usability of the components and any component improvements will be more difficult because of the created dependency. What if we keep the two components separate and create a common hook with the custom fetch functionality?

Reply:
Well, I am still a proponent of the originally suggested implementation. If you have concrete evidence that we are forcing two scenarios into one I would like to hear it but so far I don't see it yet.

Let's take a look at the scenario that the GeoJSONLayer is providing, which to me is the following:

> The GeoJSONLayer takes a URL that points to a GeoJSON document. It retrieves and displays the content from the URL and adds it as a layer to the map.

If this is indeed the scenario of the GeoJSONLayer then it should also be responsible for properly handling changes to it's dependent props such as the URL.

Do you disagree with this statement? If so I would like to know what the alternative would be in your opinion.

Reply:
I'm gonna move this conversation to a different PR as there is no need to keep this as a blocker.